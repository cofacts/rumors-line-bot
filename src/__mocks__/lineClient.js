export default {
  verifyIDToken: jest.fn(),
  revokeNotifyToken: jest.fn(),
};
