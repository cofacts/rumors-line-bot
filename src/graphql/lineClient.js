import fetch from 'node-fetch';
import querystring from 'querystring';
import rollbar from 'src/lib/rollbar';

/**
 * Reference: https://developers.line.biz/en/reference/social-api/#response-5
 *
 * @param {string} idToken
 * @return {object} Response from API
 */
export async function verifyIDToken(idToken) {
  const URL = 'https://api.line.me/oauth2/v2.1/verify';
  const resp = await fetch(URL, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
    },
    body: querystring.stringify({
      id_token: idToken,
      client_id: process.env.LINE_LOGIN_CHANNEL_ID,
    }),
  });

  const result = await resp.json();

  if (resp.status !== 200) {
    rollbar.error(`[verifyIDToken] ${resp.status}: ${result.message}.`, {
      result,
    });
  }

  return result;
}

/**
 * Reference: https://notify-bot.line.me/doc/en/
 *
 * @param {string} lineNotifyToken
 */
export async function revokeNotifyToken(lineNotifyToken) {
  const URL = 'https://notify-api.line.me/api/revoke';
  const resp = await fetch(URL, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
      Authorization: `Bearer ${lineNotifyToken}`,
    },
  });
  const result = await resp.json();

  if (resp.status !== 200) {
    rollbar.error(`[revokeNotifyToken] ${resp.status}: ${result.message}.`, {
      result,
    });
  }

  return result;
}
