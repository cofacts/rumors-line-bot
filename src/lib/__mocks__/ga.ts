const eventMock = jest.fn();
const sendMock = jest.fn();
const setMock = jest.fn();

const ga = Object.assign(
  jest.fn().mockImplementation(() => ({
    event: eventMock,
    send: sendMock,
    set: setMock,
    screenview: () => undefined,
  })),
  {
    clearAllMocks: () => {
      eventMock.mockClear();
      sendMock.mockClear();
      setMock.mockClear();
      ga.mockClear();
    },
    eventMock,
    sendMock,
    setMock,
  }
);

export default ga;
