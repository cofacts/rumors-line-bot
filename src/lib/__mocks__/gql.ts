const mockResultQueue: object[] = [];

function gqlMock() {
  return () => {
    const ret = mockResultQueue.shift();
    return Promise.resolve(ret);
  };
}

/**
 * Enqueues mock data what will be returned in next gql()() call.
 *
 * Once gql()() is invoked, the first data in queue will be consumed.
 *
 * @param {*} returnValue the mock value that gql()()'s returned promise will resolve to.
 * @returns nothing
 */
gqlMock.__push = function (returnValue: object) {
  mockResultQueue.push(returnValue);
};

/**
 * Empty the return data queue.
 *
 * Invoking this in afterEach() will ensure 1 failed test would not leave side effects to other
 * tests.
 */
gqlMock.__reset = function () {
  mockResultQueue.length = 0;
};

/**
 * @returns {Boolean} if the mock response has been depleted
 */
gqlMock.__finished = function () {
  return mockResultQueue.length === 0;
};

export default gqlMock;
export type MockedGql = typeof gqlMock;
