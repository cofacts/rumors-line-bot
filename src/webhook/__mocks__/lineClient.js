const lineClient = {
  post: jest.fn(),
  get: jest.fn(),
  getContent: jest.fn(),
};
export default lineClient;
