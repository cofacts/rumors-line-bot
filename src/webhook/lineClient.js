import fetch from 'node-fetch';
import rollbar from 'src/lib/rollbar';

export default { post, get, getContent };

async function post(endpoint = '', body = {}, options = {}) {
  const URL = `https://api.line.me/v2/bot${endpoint}`;
  const resp = await fetch(URL, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${process.env.LINE_CHANNEL_TOKEN}`,
      ...options.headers,
    },
    ...options,
    body: JSON.stringify(body),
  });

  const result = await resp.json();

  if (resp.status !== 200) {
    console.error(JSON.stringify(result, null, '  '));

    rollbar.error(
      `[LINE Client] ${resp.status}: ${result.message}.`,
      {
        // Request object for rollbar server SDK
        headers: options.headers,
        body: JSON.stringify(body),
        url: URL,
        method: 'POST',
      },
      { result }
    );
  }

  return result;
}

async function get(endpoint = '', options = {}) {
  const URL = `https://api.line.me/v2/bot${endpoint}`;
  const resp = await fetch(URL, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${process.env.LINE_CHANNEL_TOKEN}`,
      ...options.headers,
    },
    ...options,
  });

  const result = await resp.json();

  if (resp.status !== 200) {
    console.error(JSON.stringify(result, null, '  '));

    rollbar.error(
      `[LINE Client] ${resp.status}: ${result.message}.`,
      {
        // Request object for rollbar server SDK
        headers: options.headers,
        url: URL,
        method: 'GET',
      },
      { result }
    );
  }

  return result;
}

async function getContent(messageId, options = {}) {
  // this endpoint is for sending and receiving large amounts of data in the LINE platform for Messaging API.
  // https://developers.line.biz/en/reference/messaging-api/#get-content
  const URL = `https://api-data.line.me/v2/bot/message/${messageId}/content`;
  const resp = await fetch(URL, {
    method: 'GET',
    headers: {
      Authorization: `Bearer ${process.env.LINE_CHANNEL_TOKEN}`,
      ...options.headers,
    },
    ...options,
  });

  if (resp.status !== 200) {
    const err = await resp.json();
    console.error(JSON.stringify(err, null, '  '));

    rollbar.error(
      `[LINE Client] ${resp.status}: ${err.message}.`,
      {
        // Request object for rollbar server SDK
        headers: options.headers,
        url: URL,
        method: 'GET',
      },
      { result: err }
    );
  }

  return resp;
}
