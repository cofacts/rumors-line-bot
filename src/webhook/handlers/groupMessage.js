// spec: https://github.com/cofacts/rumors-line-bot/issues/13
import stringSimilarity from 'string-similarity';
import { t } from 'ttag';
import gql from 'src/lib/gql';
import { createGroupReplyMessages } from './utils';
import ga from 'src/lib/ga';
import lineClient from 'src/webhook/lineClient';

const SIMILARITY_THRESHOLD = 0.8;
const INTRO_KEYWORDS = ['hi cofacts', 'hi confacts'];
const LEAVE_KEYWORD = 'bye bye cofacts';
const VALID_CATEGORIES = [
  'medical', //'疾病、醫藥🆕',
  'covid19', //'COVID-19 疫情🆕',
  'mz2n7nEBrIRcahlYnQpz', //'科技、資安、隱私',
  'lT3h7XEBrIRcahlYugqq', //'保健秘訣、食品安全',
  'nD2n7nEBrIRcahlYwQoW', //'免費訊息詐騙',
];

export default async function processText(event, groupId) {
  let replies;
  if (event.input === undefined) {
    throw new Error('input undefined');
  }

  // Track text message type send by user
  const visitor = ga(groupId, '__INIT__', event.input, event.source.type);

  const inputSimilarityWithIntro = Math.max(
    ...INTRO_KEYWORDS.map(keyword => {
      return stringSimilarity.compareTwoStrings(
        event.input.toLowerCase(),
        keyword
      );
    })
  );

  if (inputSimilarityWithIntro > 0.9) {
    replies = [
      {
        type: 'text',
        text: `${t`Hi i am cofacts chat bot`}😊 `,
      },
    ];
    visitor.event({
      ec: 'UserInput',
      ea: 'Intro',
      el: '',
    });
    visitor.send();
    return { event, groupId, replies };
  }

  if (event.input.toLowerCase() === LEAVE_KEYWORD) {
    await lineClient.post(`/${event.source.type}/${groupId}/leave`);
    return { event, groupId, replies };
  }

  // skip
  if (event.input.length <= 10) {
    return { event, groupId, replies: undefined };
  }

  // Search for articles
  const {
    data: { ListArticles },
  } = await gql`
    query($text: String!) {
      ListArticles(
        filter: { moreLikeThis: { like: $text } }
        orderBy: [{ _score: DESC }]
        first: 4
      ) {
        edges {
          node {
            text
            id
            articleCategories(status: NORMAL) {
              categoryId
              positiveFeedbackCount
              negativeFeedbackCount
            }
            articleReplies(status: NORMAL) {
              reply {
                id
                type
                text
                reference
              }
              positiveFeedbackCount
              negativeFeedbackCount
            }
          }
        }
      }
    }
  `({
    text: event.input,
  });

  if (ListArticles.edges.length) {
    // Track if find similar Articles in DB.
    visitor.event({ ec: 'UserInput', ea: 'ArticleSearch', el: 'ArticleFound' });

    // Track which Article is searched. And set tracking event as non-interactionHit.
    ListArticles.edges.forEach(edge => {
      visitor.event({
        ec: 'Article',
        ea: 'Search',
        el: edge.node.id,
        ni: true,
      });
    });

    const edgesSortedWithSimilarity = ListArticles.edges
      .map(edge => {
        edge.similarity = stringSimilarity.compareTwoStrings(
          // Remove spaces so that we count word's similarities only
          //
          edge.node.text.replace(/\s/g, ''),
          event.input.replace(/\s/g, '')
        );
        return edge;
      })
      .sort((edge1, edge2) => edge2.similarity - edge1.similarity);

    const hasIdenticalDocs =
      edgesSortedWithSimilarity[0].similarity >= SIMILARITY_THRESHOLD;
    const hasValidCategory = edgesSortedWithSimilarity[0].node.articleCategories.reduce(
      (acc, articleCategory) =>
        (acc =
          acc ||
          (VALID_CATEGORIES.includes(articleCategory.categoryId) &&
            articleCategory.positiveFeedbackCount >=
              articleCategory.negativeFeedbackCount)),
      false
    );

    if (hasIdenticalDocs) {
      const node = edgesSortedWithSimilarity[0].node;
      visitor.event({
        ec: 'Article',
        ea: 'Selected',
        el: node.id,
      });

      if (hasValidCategory) {
        const articleReply = getValidArticleReply(node);
        if (articleReply) {
          visitor.event({
            ec: 'Reply',
            ea: 'Selected',
            el: articleReply.reply.id,
          });

          replies = createGroupReplyMessages(
            event.input,
            articleReply.reply,
            node.articleReplies.length,
            node.id
          );
        }
      }
    }

    console.debug(
      '[GROUP]',
      JSON.stringify({
        similarities: edgesSortedWithSimilarity.map(({ similarity }) =>
          similarity.toFixed(2)
        ),
        hasIdenticalDocs,
        hasValidCategory,
        hasReply: !!replies,
      })
    );

    // sends ga only when article found
    visitor.send();
  }
  //else {
  // Article not found
  // }

  return { event, groupId, replies };
}

/**
 * Return a article reply which type is rumor.
 *
 * And to make sure there's no controversy,
 * It should
 * 1. check if rumor count >> non-rumor count when replyCount > 2
 * 2. candidate's positiveFeedbackCount > candidate's negativeFeedbackCount
 * 3. candidate's positiveFeedbackCount > non-rumors' positiveFeedbackCount
 *
 * @param {object} articleReplies `Article.articleReplies` from rumors-api
 * @returns {object} A reply which type is rumor
 */
export function getValidArticleReply({ articleReplies }) {
  let rumorCount = 0;
  const replyCount = articleReplies.length;
  const postiveArticleReplies = articleReplies
    .reduce((result, ar) => {
      if (ar.positiveFeedbackCount > ar.negativeFeedbackCount) {
        result.push(ar);
      }
      if (ar.reply.type === 'RUMOR') rumorCount++;
      return result;
    }, [])
    .sort(
      (reply1, reply2) =>
        reply2.positiveFeedbackCount - reply1.positiveFeedbackCount
    );

  // if replyCount > 2, check if rumorCount >= 2/3 replyCount
  if (replyCount <= 2 || rumorCount >= (replyCount * 2) / 3) {
    const candidate = postiveArticleReplies[0];

    // return `undefined` if first reply's type is not rumor
    if (!candidate || candidate.reply.type !== 'RUMOR') return undefined;

    // get first non-rumor articleReply (it may be undefined)
    const nonRumorAR = postiveArticleReplies.find(
      ar => ar.reply.type !== 'RUMOR'
    );

    // return candidate if its positiveFeedbackCount > first non-rumor articleReply
    if (
      nonRumorAR === undefined ||
      candidate.positiveFeedbackCount > nonRumorAR.positiveFeedbackCount
    )
      return candidate;
  }
  return undefined;
}
