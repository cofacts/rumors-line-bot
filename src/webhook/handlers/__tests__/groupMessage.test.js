jest.mock('src/lib/ga');
jest.mock('src/lib/gql');
jest.mock('src/webhook/lineClient');

import ga from 'src/lib/ga';
import gql from 'src/lib/gql';

import groupMessage, { getValidArticleReply } from '../groupMessage';

import {
  apiResult,
  article as articleFixtures,
} from '../__fixtures__/groupMessage';
import lineClient from 'src/webhook/lineClient';

const event = {
  groupId: 'C904bb9fc2f4904b2facf8204b3f08c79',
  source: { type: 'group' },
};

beforeEach(() => {
  event.input = undefined;
  ga.clearAllMocks();
  gql.__reset();
  lineClient.post.mockClear();
});

describe('groupMessage', () => {
  it('rejects undefined input', async () => {
    await expect(groupMessage(event)).rejects.toMatchInlineSnapshot(
      `[Error: input undefined]`
    );
  });

  it('should not process input length less then 11', async () => {
    event.input = '0123456789';
    gql.__push(apiResult.validArticleAndOneReply);
    expect((await groupMessage(event)).replies).toBeUndefined();
    expect(gql.__finished()).toBe(false);
    expect(ga.sendMock).toHaveBeenCalledTimes(0);
  });

  it('processes introduction message Hi Cofacts', async () => {
    gql.__push(apiResult.notFound);
    event.input = 'Hi Cofacts';
    let result = await groupMessage(event);
    expect(result.replies).not.toBeUndefined();
    expect(result).toMatchSnapshot();
    event.input = 'hi confacts';
    result = await groupMessage(event);
    expect(result.replies).not.toBeUndefined();
    expect(result).toMatchSnapshot();
    expect(gql.__finished()).toBe(false);

    // don't really care about this result :p
    event.input = 'cofacts';
    expect(await groupMessage(event)).toMatchSnapshot();
  });

  it('processes leave command bye bye cofacts', async () => {
    gql.__push(apiResult.notFound);
    event.input = 'bye bye cofacts';
    let result = await groupMessage(event);
    expect(gql.__finished()).toBe(false);
    expect(result.replies).toBeUndefined();
    expect(lineClient.post).toHaveBeenCalledTimes(1);
  });

  it('should not reply and send ga if article not found', async () => {
    event.input = 'article_not_found';
    gql.__push(apiResult.notFound);
    expect((await groupMessage(event)).replies).toBeUndefined();
    expect(gql.__finished()).toBe(true);
    expect(ga.sendMock).toHaveBeenCalledTimes(0);
  });

  it('should handle valid article with one category', async () => {
    event.input =
      'WHO 最新研究顯示 Covid-19 其實源自黑暗料理界，即日起正名為「黑料病毒」';
    gql.__push(apiResult.validArticleWithOneCategory);
    const result = await groupMessage(event);
    expect(result.replies).not.toBeUndefined();
    expect(result).toMatchSnapshot();
    expect(gql.__finished()).toBe(true);
    expect(ga.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          undefined,
          "__INIT__",
          "WHO 最新研究顯示 Covid-19 其實源自黑暗料理界，即日起正名為「黑料病毒」",
          "group",
        ],
      ]
    `);
    expect(ga.eventMock.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          Object {
            "ea": "ArticleSearch",
            "ec": "UserInput",
            "el": "ArticleFound",
          },
        ],
        Array [
          Object {
            "ea": "Search",
            "ec": "Article",
            "el": "3nbzf064ks60d",
            "ni": true,
          },
        ],
        Array [
          Object {
            "ea": "Selected",
            "ec": "Article",
            "el": "3nbzf064ks60d",
          },
        ],
        Array [
          Object {
            "ea": "Selected",
            "ec": "Reply",
            "el": "fake-reply-id",
          },
        ],
      ]
    `);
    expect(ga.sendMock).toHaveBeenCalledTimes(1);
  });

  it('should handle valid article with two categories', async () => {
    event.input = '你知道黑啤愛吃什麼嗎？ 黑啤愛吃蠶寶寶！';
    gql.__push(apiResult.validArticleWithTwoCategories);
    const result = await groupMessage(event);
    expect(result.replies).not.toBeUndefined();
    expect(result).toMatchSnapshot();
    expect(gql.__finished()).toBe(true);
    expect(ga.eventMock.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          Object {
            "ea": "ArticleSearch",
            "ec": "UserInput",
            "el": "ArticleFound",
          },
        ],
        Array [
          Object {
            "ea": "Search",
            "ec": "Article",
            "el": "3nbzf064ks60d",
            "ni": true,
          },
        ],
        Array [
          Object {
            "ea": "Search",
            "ec": "Article",
            "el": "8nbzf064ks87g",
            "ni": true,
          },
        ],
        Array [
          Object {
            "ea": "Search",
            "ec": "Article",
            "el": "2zn1215x6e70v",
            "ni": true,
          },
        ],
        Array [
          Object {
            "ea": "Selected",
            "ec": "Article",
            "el": "3nbzf064ks60d",
          },
        ],
        Array [
          Object {
            "ea": "Selected",
            "ec": "Reply",
            "el": "fake-reply-id1",
          },
        ],
      ]
    `);
    expect(ga.sendMock).toHaveBeenCalledTimes(1);
  });

  it('should handle invalid article category feedback', async () => {
    event.input = '你知道嗎？其實黑啤愛吃蠶寶寶哦！';
    gql.__push(apiResult.invalidCategoryFeedback);
    expect((await groupMessage(event)).replies).toBeUndefined();
    expect(gql.__finished()).toBe(true);
    expect(ga.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          undefined,
          "__INIT__",
          "你知道嗎？其實黑啤愛吃蠶寶寶哦！",
          "group",
        ],
      ]
    `);
    expect(ga.eventMock.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          Object {
            "ea": "ArticleSearch",
            "ec": "UserInput",
            "el": "ArticleFound",
          },
        ],
        Array [
          Object {
            "ea": "Search",
            "ec": "Article",
            "el": "3nbzf064ks60d",
            "ni": true,
          },
        ],
        Array [
          Object {
            "ea": "Selected",
            "ec": "Article",
            "el": "3nbzf064ks60d",
          },
        ],
      ]
    `);
    expect(ga.sendMock).toHaveBeenCalledTimes(1);
  });

  it('should handle invalid article category', async () => {
    event.input = '以後吃蘋果一定要削皮。';
    gql.__push(apiResult.invalidArticleCategory);
    expect((await groupMessage(event)).replies).toBeUndefined();
    expect(gql.__finished()).toBe(true);
    expect(ga.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          undefined,
          "__INIT__",
          "以後吃蘋果一定要削皮。",
          "group",
        ],
      ]
    `);
    expect(ga.eventMock.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          Object {
            "ea": "ArticleSearch",
            "ec": "UserInput",
            "el": "ArticleFound",
          },
        ],
        Array [
          Object {
            "ea": "Search",
            "ec": "Article",
            "el": "2zn1215x6e70v",
            "ni": true,
          },
        ],
        Array [
          Object {
            "ea": "Selected",
            "ec": "Article",
            "el": "2zn1215x6e70v",
          },
        ],
      ]
    `);
    expect(ga.sendMock).toHaveBeenCalledTimes(1);
  });

  it('should handle input is not identical to article ', async () => {
    event.input = '我知道黑啤愛吃蛾哦！';
    gql.__push(apiResult.invalidArticleReply);
    expect((await groupMessage(event)).replies).toBeUndefined();
    expect(gql.__finished()).toBe(true);
    expect(ga.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          undefined,
          "__INIT__",
          "我知道黑啤愛吃蛾哦！",
          "group",
        ],
      ]
    `);
    expect(ga.eventMock.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          Object {
            "ea": "ArticleSearch",
            "ec": "UserInput",
            "el": "ArticleFound",
          },
        ],
        Array [
          Object {
            "ea": "Search",
            "ec": "Article",
            "el": "3nbzf064ks60d",
            "ni": true,
          },
        ],
      ]
    `);
    expect(ga.sendMock).toHaveBeenCalledTimes(1);
  });

  it('should handle invalid article reply', async () => {
    event.input = '我不會說我知道黑啤愛吃蠶寶寶哦！';
    gql.__push(apiResult.invalidArticleReply);
    expect((await groupMessage(event)).replies).toBeUndefined();
    expect(gql.__finished()).toBe(true);
    expect(ga.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          undefined,
          "__INIT__",
          "我不會說我知道黑啤愛吃蠶寶寶哦！",
          "group",
        ],
      ]
    `);
    expect(ga.eventMock.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          Object {
            "ea": "ArticleSearch",
            "ec": "UserInput",
            "el": "ArticleFound",
          },
        ],
        Array [
          Object {
            "ea": "Search",
            "ec": "Article",
            "el": "3nbzf064ks60d",
            "ni": true,
          },
        ],
        Array [
          Object {
            "ea": "Selected",
            "ec": "Article",
            "el": "3nbzf064ks60d",
          },
        ],
      ]
    `);
    expect(ga.sendMock).toHaveBeenCalledTimes(1);
  });
});

describe('processes articleReplies which article is valid', () => {
  it('invalid reply type', () => {
    expect(
      getValidArticleReply(articleFixtures.invalidReplyType)
    ).toBeUndefined();
  });

  it('invalid reply feedback count', () => {
    expect(
      getValidArticleReply(articleFixtures.invalidReplyFeedbackCount)
    ).toBeUndefined();
  });

  it('invalidTwoReplies1', () => {
    expect(
      getValidArticleReply(articleFixtures.invalidTwoReplies1)
    ).toBeUndefined();
  });

  it('invalidTwoReplies2', () => {
    expect(
      getValidArticleReply(articleFixtures.invalidTwoReplies2)
    ).toBeUndefined();
  });

  it('invalidThreeReplies1', () => {
    expect(
      getValidArticleReply(articleFixtures.invalidThreeReplies1)
    ).toBeUndefined();
  });

  it('invalidThreeReplies2', () => {
    expect(
      getValidArticleReply(articleFixtures.invalidThreeReplies2)
    ).toBeUndefined();
  });

  it('invalidMultipleReplies1', () => {
    expect(
      getValidArticleReply(articleFixtures.invalidMultipleReplies1)
    ).toBeUndefined();
  });
  it('invalidMultipleReplies2', () => {
    expect(
      getValidArticleReply(articleFixtures.invalidMultipleReplies2)
    ).toBeUndefined();
  });
  it('invalidMultipleReplies3', () => {
    expect(
      getValidArticleReply(articleFixtures.invalidMultipleReplies3)
    ).toBeUndefined();
  });

  it('twoReplies1 RUMOR positiveFeedbackCount > Non-RUMOR', () => {
    expect(getValidArticleReply(articleFixtures.twoReplies1)).toMatchSnapshot();
  });

  it('twoReplies2 equal positiveFeedbackCount, but both type are RUMOR', () => {
    expect(getValidArticleReply(articleFixtures.twoReplies2)).toMatchSnapshot();
  });

  it('threeReplies1 should return reply has more positiveFeedbackCount', () => {
    expect(
      getValidArticleReply(articleFixtures.threeReplies1)
    ).toMatchSnapshot();
  });

  it('threeReplies2 should return reply has second highest positiveFeedbackCount', () => {
    expect(
      getValidArticleReply(articleFixtures.threeReplies2)
    ).toMatchSnapshot();
  });

  it('multipleReplies1 should return Rumor reply1 or Rumor reply3', () => {
    expect(
      getValidArticleReply(articleFixtures.multipleReplies1)
    ).toMatchSnapshot();
  });

  it('multipleReplies2 should return reply has second highest positiveFeedbackCount', () => {
    expect(
      getValidArticleReply(articleFixtures.multipleReplies2)
    ).toMatchSnapshot();
  });
});
