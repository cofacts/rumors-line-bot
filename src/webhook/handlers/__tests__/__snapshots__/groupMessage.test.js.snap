// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`groupMessage processes introduction message Hi Cofacts 1`] = `
Object {
  "event": Object {
    "groupId": "C904bb9fc2f4904b2facf8204b3f08c79",
    "input": "Hi Cofacts",
    "source": Object {
      "type": "group",
    },
  },
  "groupId": undefined,
  "replies": Array [
    Object {
      "text": "Hi i am cofacts chat bot😊 ",
      "type": "text",
    },
  ],
}
`;

exports[`groupMessage processes introduction message Hi Cofacts 2`] = `
Object {
  "event": Object {
    "groupId": "C904bb9fc2f4904b2facf8204b3f08c79",
    "input": "hi confacts",
    "source": Object {
      "type": "group",
    },
  },
  "groupId": undefined,
  "replies": Array [
    Object {
      "text": "Hi i am cofacts chat bot😊 ",
      "type": "text",
    },
  ],
}
`;

exports[`groupMessage processes introduction message Hi Cofacts 3`] = `
Object {
  "event": Object {
    "groupId": "C904bb9fc2f4904b2facf8204b3f08c79",
    "input": "cofacts",
    "source": Object {
      "type": "group",
    },
  },
  "groupId": undefined,
  "replies": undefined,
}
`;

exports[`groupMessage should handle valid article with one category 1`] = `
Object {
  "event": Object {
    "groupId": "C904bb9fc2f4904b2facf8204b3f08c79",
    "input": "WHO 最新研究顯示 Covid-19 其實源自黑暗料理界，即日起正名為「黑料病毒」",
    "source": Object {
      "type": "group",
    },
  },
  "groupId": undefined,
  "replies": Array [
    Object {
      "text": "Thank you for sharing “WHO 最新研究顯示⋯⋯”😊 
I found that there are some disagreement to the message:",
      "type": "text",
    },
    Object {
      "text": "It\`s rumor. It\`s rumor. It\`s rumor.",
      "type": "text",
    },
    Object {
      "text": "references：www.who.int",
      "type": "text",
    },
    Object {
      "text": "⬆️ Therefore, the author think the message contains misinformation.

💁 This content is provided by Cofact message reporting chatbot and crowd-sourced fact-checking community under CC BY-SA 4.0 license. Please refer to their references and make judgements on your own.


⁉️ If you have different thoughts, you may have your say here:
https://dev.cofacts.tw/article/3nbzf064ks60d",
      "type": "text",
    },
  ],
}
`;

exports[`groupMessage should handle valid article with two categories 1`] = `
Object {
  "event": Object {
    "groupId": "C904bb9fc2f4904b2facf8204b3f08c79",
    "input": "你知道黑啤愛吃什麼嗎？ 黑啤愛吃蠶寶寶！",
    "source": Object {
      "type": "group",
    },
  },
  "groupId": undefined,
  "replies": Array [
    Object {
      "text": "Thank you for sharing “你知道黑啤愛吃什麼嗎⋯⋯”😊 
I found that there are some disagreement to the message:",
      "type": "text",
    },
    Object {
      "text": "這是謠言！這是謠言！這是謠言！這是謠言！",
      "type": "text",
    },
    Object {
      "text": "references：https://taiwanbar.cc/",
      "type": "text",
    },
    Object {
      "text": "⬆️ Therefore, the author think the message contains misinformation.

💁 This content is provided by Cofact message reporting chatbot and crowd-sourced fact-checking community under CC BY-SA 4.0 license. Please refer to their references and make judgements on your own.


⁉️ If you have different thoughts, you may have your say here:
https://dev.cofacts.tw/article/3nbzf064ks60d",
      "type": "text",
    },
  ],
}
`;

exports[`processes articleReplies which article is valid multipleReplies1 should return Rumor reply1 or Rumor reply3 1`] = `
Object {
  "negativeFeedbackCount": 0,
  "positiveFeedbackCount": 2,
  "reply": Object {
    "text": "Rumor reply1 has second highest positiveFeedbackCount and equals to reply3",
    "type": "RUMOR",
  },
}
`;

exports[`processes articleReplies which article is valid multipleReplies2 should return reply has second highest positiveFeedbackCount 1`] = `
Object {
  "negativeFeedbackCount": 2,
  "positiveFeedbackCount": 3,
  "reply": Object {
    "text": "Rumor reply3 has second highest positiveFeedbackCount",
    "type": "RUMOR",
  },
}
`;

exports[`processes articleReplies which article is valid threeReplies1 should return reply has more positiveFeedbackCount 1`] = `
Object {
  "negativeFeedbackCount": 0,
  "positiveFeedbackCount": 2,
  "reply": Object {
    "text": "Rumor reply1 has more positiveFeedbackCount",
    "type": "RUMOR",
  },
}
`;

exports[`processes articleReplies which article is valid threeReplies2 should return reply has second highest positiveFeedbackCount 1`] = `
Object {
  "negativeFeedbackCount": 0,
  "positiveFeedbackCount": 2,
  "reply": Object {
    "text": "Rumor reply1 has second highest positiveFeedbackCount",
    "type": "RUMOR",
  },
}
`;

exports[`processes articleReplies which article is valid twoReplies1 RUMOR positiveFeedbackCount > Non-RUMOR 1`] = `
Object {
  "negativeFeedbackCount": 0,
  "positiveFeedbackCount": 2,
  "reply": Object {
    "text": "Rumor reply",
    "type": "RUMOR",
  },
}
`;

exports[`processes articleReplies which article is valid twoReplies2 equal positiveFeedbackCount, but both type are RUMOR 1`] = `
Object {
  "negativeFeedbackCount": 0,
  "positiveFeedbackCount": 2,
  "reply": Object {
    "text": "Rumor reply1",
    "type": "RUMOR",
  },
}
`;
